### ISSUE TITLE
MQTT publish fails in Online Mode and data is lost

### ISSUE THREAD

**Issue by LeSuedois** [LeClubSandwichStudio/LaCOOLBoard#12]
This issue concerns the weather station example or any other CoolBoard with the sleep flag active. thus this issue get's highest priority for me.
When the publication fails the board dumps the data and goes to sleep. the wifi is connected but something goes wrong and the publish returns a fail. instead of retrying to push or save the data to json the card goes ahead, dumps the data and goes to sleep..

**Comment by LeSuedois** [comment]
Ok, it's better, but the issue is not solved. For instance we save the metrics as JSON to send them later. This does not interfere with OffLine Mode.
At least we don't loose the data...
Work in Progress...
commit 526f69460d78399542d89d27ac75366e50b2c133

**Comment by LeSuedois** [comment]
I spend the day on this issue and I think I got it working. In first instance It looked like a strange WiFi problem but finally I found out that the MQTT connection is either to slow or just looses connection to the server or whatever. 
I reinforced the publish function and some stuff in CoolBoard.cpp to ensure that no data is lost with lousy internet performances. 
This must be stress tested over the next days..
commit f4525ee52b4eeb1bd6f0e8aae0d5c3a84d4eeb31

**Comment by LeSuedois** [comment]
After some searching I recognized that the user name in mqttConfig was used as name for the mqtt client. since I MYSELF I named all shipped card's "my first CoolBoard", the server got multiple clients with the same name (wich IMHO is forbidden in the MQTT protocol) and the server did his job and kicked them out. 
I worked on this issue for a weak. first we reinforced all connections retry's regarding WiFi and MQTT, wich is not bad at all..
well, is seems to work nice now. the branch was merged, but I still let the issue open until the firmware is deployed on more coolboards..
https://github.com/LaCoolCo/LaCOOLBoard/commit/03b514314e0fd63a725e2439543a4b0883aade5a
